sockeye/inference.py

```diff
@@ -511,35 +529,145 @@ def get_max_output_length(input_length: int):
 
 
 Tokens = List[str]
-TranslatorInput = NamedTuple('TranslatorInput', [
-    ('id', int),
-    ('sentence', str),
-    ('tokens', Tokens),
-])
-"""
-Required input for Translator.
 
-:param id: Sentence id.
-:param sentence: Input sentence.
-:param tokens: List of input tokens.
-"""
 
-InputChunk = NamedTuple("InputChunk",
-                        [("id", int),
-                         ("chunk_id", int),
-                         ("tokens", Tokens)])
-"""
-A chunk of a TranslatorInput.
+class TranslatorInput:
+    """
+    Object required by Translator.translate().
+
+    :param sentence_id: Sentence id.
+    :param tokens: List of input tokens.
+    :param factors: Optional list of additional factor sequences.
+    :param chunk_id: Chunk id. Defaults to -1.
+    """
+
+    def __init__(self,
+                 sentence_id: int,
+                 tokens: Tokens,
+                 factors: Optional[List[Tokens]] = None,
+                 chunk_id: int = -1) -> None:
+        self.sentence_id = sentence_id
+        self.chunk_id = chunk_id
+        self.tokens = tokens
+        self.factors = factors
+
+    def __str__(self):
+        return 'TranslatorInput(%d, %s, %s, %d)' % (self.sentence_id, self.tokens, self.factors, self.chunk_id)
+
+    def __len__(self):
+        return len(self.tokens)
+
+    @property
+    def num_factors(self) -> int:
+        """
+        Returns the number of factors of this instance.
+        """
+        return 1 + (0 if not self.factors else len(self.factors))
+
+    def chunks(self, chunk_size: int) -> Generator['TranslatorInput', None, None]:
+        """
+        Takes a TranslatorInput (itself) and yields TranslatorInputs for chunks of size chunk_size.
+
+        :param chunk_size: The maximum size of a chunk.
+        :return: A generator of TranslatorInputs, one for each chunk created.
+        """
+        for chunk_id, i in enumerate(range(0, len(self), chunk_size)):
+            factors = [factor[i:i + chunk_size] for factor in self.factors] if self.factors is not None else None
+            yield TranslatorInput(sentence_id=self.sentence_id,
+                                  tokens=self.tokens[i:i + chunk_size],
+                                  factors=factors,
+                                  chunk_id=chunk_id)
+
+
+def make_input_from_plain_string(sentence_id: int, string: str) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a plain string.
+
+    :param sentence_id: An integer id.
+    :param string: An input string.
+    :return: A TranslatorInput.
+    """
+    return TranslatorInput(sentence_id, tokens=list(data_io.get_tokens(string)), factors=None)
+
+
+def make_input_from_json_string(sentence_id: int, json_string: str) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a JSON object, serialized as a string.
+
+    :param sentence_id: An integer id.
+    :param json_string: A JSON object serialized as a string.
```

Comment:
It might make sense to briefly mention the format in which the data is expected to be? Namely, the structure where there is one entry per factor (rather than something like a list of tuples `[("token", "factor1", ...)])`. etc

Reply:
done